t/templates.t

```diff
@@ -0,0 +1,246 @@
+    'template defs: read all template sets';
+
+###############################
+# Template definitions errors #
+###############################
+
+throws_ok {
+        my $templates_defs = App::DuckPAN::TemplateDefinitions->new(
+            templates_yml => 't/template/templates-nonexistent.yml'
+        );
+    } qr/Error loading/,
+    'template defs: non-existent template definitions file throws error';
+
+########################
+# Individual Templates #
+########################
+
+# check if all fields of a template are read
+is $template_map{pm}->label, 'Perl Module',
+    'template defs: set template label field';
+is $template_map{pm}->input_file, 't/template/lib/DDG/Default.pm',
+    'template defs: set template input field';
+is $template_map{pm}->output_file, 't/out/lib/DDG/<:$package_name:>.pm',
+    'template defs: set template output field';
+ok $template_map{pm}->needs_restart,
+    'template defs: set template needs_restart field';
+
+# output directory computation
+is $template_map{pm}->output_directory, 't/out/lib/DDG',
+    'template: output directory generated';
+is $template_map{complex_out_dir}->output_directory, 't/out/share/text',
+    'template: output directory generated (complex)';
+
+#################
+# Template Sets #
+#################
+
+# check if all fields of a template set are read
+my $template_set_ro = $template_set_map{required_and_optional};
+
+is $template_set_ro->description, 'Template set with required and optional templates',
+    'template defs: set template_set description';
+is_deeply $template_set_ro->required_templates, [@template_map{qw(pm test)}],
+    'template defs: set template_set required templates';
+is_deeply $template_set_ro->optional_templates, [@template_map{qw(js css)}],
+    'template defs: set template_set optional templates';
+
+ok !$template_set_map{subdir_support_specified}->subdir_support, 'Template set with subdir support explicitly disabled';
+ok $template_set_map{subdir_support_not_defined}->subdir_support, 'Template set with subdir support implicitly enabled';
+
+{
+    # check template combinations that we should show to the user
+    my $combinations = $template_set_map{all_optional}->optional_template_combinations;
+
+    # for neater code
+    my %tm = %template_map;
+
+    # combinations must be in a particular order
+    is_deeply $combinations,
+        [
+            [ $tm{js}   ],
+            [ $tm{pm}   ],
+            [ $tm{test} ],
+            [ $tm{js},  $tm{pm}   ],
+            [ $tm{js},  $tm{test} ],
+            [ $tm{pm},  $tm{test} ],
+            [ $tm{js},  $tm{pm},  $tm{test} ],
+        ],
+        'template set: all optional template combinations generated correctly';
+}
+
+##################################
+# File generation from templates #
+##################################
+
+my $package_name = 'MyInstantAnswer';
+my $lia_name = 'my_instant_answer';
```

Comment:
@srvsh I think we need to remove reference to this now? Or change it to `ia_id?`